Edit IIIF bitstream metadata

mspalti opened the pull request:
## References
* Fixes #1388

## Description
Adds the ability to edit IIIF bitstream fields. The additional IIIF fields appear in the bitstream edit form when the parent item is IIIF-enabled.

## Instructions for Reviewers
Please add a more detailed description of the changes made by your PR. At a minimum, providing a bulleted list of changes in your PR is helpful to reviewers.

List of changes in this PR:
* The bitstream edit component has 4 additional IIIF fields
* The parent item is checked so that the new fields appear only when iiif support is enabled
* The bitstream object is updated with IIIF metadata on submission

**Include guidance for how to test or review your PR.** 
When you edit a bitstream  that is attached to an item with `dspace.iiif.enabled` metadata (`value = true`) you should see the four additional fields. It should be possible to successfully update the metadata. To see results in the mirador viewer, use `yarn run build:mirador` and `yarn start`. 

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

mspalti commented:
I made a few additions to this code after testing. Before displaying IIIF fields in the form, the bitstream mimetype and the bundle are checked. Previously only the item was checked to see if IIIF was enabled. This prevents the IIIF form options from appearing where they aren't needed.

I also modified the bitstream metadata update so that if a form element is empty the field is removed from the bitstream metadata. Some of our logic on the REST side was triggered by the existence of a field regardless of it's content, which could be an empty string or null. This change allows the user to remove a metadata field from the object by removing the existing value.

mspalti commented on src/app/bitstream-page/edit-bitstream-page/edit-bitstream-page.component.ts:
```diff
@@ -497,6 +627,68 @@ export class EditBitstreamPageComponent implements OnInit, OnDestroy {
+              }
+            });
+            // Assure that the form always detects the iiif addition.
+            this.changeDetectorRef.detectChanges();
```
I'm not sure this is necessary...

tdonohue commented on src/assets/i18n/en.json5:
```diff
@@ -582,6 +582,23 @@
 
+  "bitstream.edit.form.iiifLabel.label": "IIIF Label",
+
+  "bitstream.edit.form.iiifLabel.hint": "Canvas label for this image",
```
The hints never are displayed in the form.

mspalti replied:
It appears that hints weren't working because the component was using the `DynamicInputModel` constructor and not `DsDynamicInputModel` which extends it. Hints now display for iiif fields and for the file name and description fields.

The spacing issue fixed by adding a top-margin to the "label" in `ds-dynamic-form-control-container.component.scss`. This is is going to affect any component that relies on the container. I tried to find another usage but could not. Still, adding a bit of padding the the label seems like something that's always helpful...

label {
  margin-top: calc(0.7 * var(--ds-content-spacing));
}

Canvas width is now updating.

tdonohue reviewed (changes requested):
@mspalti : Apologies for the delays in testing...this one accidentally got forgotten. In any case, I tested this today, and it mostly seems to work.  However, I've hit a few small issues:

1. The "Canvas width" field doesn't work for me at all.  All other fields save their values, but I cannot get the "width" to be saved.
2. The new IIIF field labels are slightly misaligned. They almost bump up against the field above them, which makes the form harder to read.  Here's a screenshot...notice the (good) alignment of the "Selected Format" field, while the "IIIF" fields all look like they are bumping up against the textbox above them. 
![misaligned](https://user-images.githubusercontent.com/483997/149975007-48a0081a-007d-48a8-bd17-a2322a94cb80.png)
3. Finally, none of your hints appear in the form.  I see you created hint text, but somehow it's not displayed.  It'd be good to display it if possible, as these IIIF fields may not be self explanatory unless you know IIIF well.

Beyond that, the updates work.  I've given the code a quick glance, and it looks reasonable, though I'm hoping @atarix83 can give it a code review from his perspective (as he may be able to provide better Angular advice, as necessary).

tdonohue commented:
@mspalti : Thanks for the updates, it looks like this is failing builds now cause of minor `lint` errors.  Here's what it's reporting:
```
/home/runner/work/dspace-angular/dspace-angular/src/app/bitstream-page/edit-bitstream-page/edit-bitstream-page.component.ts:115:46
ERROR: 115:46   semicolon    Missing semicolon
ERROR: 125:37   semicolon    Missing semicolon
```

mspalti commented:
So this component hasn't been working as intended. The labels and hints are initialized using an `updateFieldTranslations()` method that looks great. But the result is that labels appear in the UI but hints are not even though they do actually exist on the object.

Both labels and hints appear when I initialize the form model with `DsDynamicInputModel` (etc) instead of Angular's `DynamicInputModel`. That's a little surprising since when I write to console the "hint" does exist on the model. Note that "hint" is a property of  `DsDynamicInputModel` and not a property of `DynamicInputModel` so it could be that Angular is not aware of the hint property for some reason.

In other DSpace forms the `DynamicInputModel` is used and hints are added successfully. But the components that I've seen do work differently in that they get the translations first and initialize the `DynamicInputModel` inside a subscription.  I think it's  a good idea to get this form working in the same way since I did notice one layout oddity when mixing types inside the same form group.

mspalti commented:
While I was looking at forms to see if hints were working I also looked at the label spacing. The margin that I added to the label isn't needed elsewhere. So I removed it. There seems to be something up with this particular form layout.

mspalti commented:
OK, @tdonohue and @atarix83 , when you have time this is ready for review. I've managed to address the feedback in a way that's hopefully good enough for now. 

- Spacing was incorrect because `ds-dynamic-form-control-container > div` didn't have the `form-row` class that's included in other forms. (This was affecting everything on the edit bitstream form, not just the new iiif form fields.) I compensated by adding a new rule to `edit-bitstream-page.component.scss`.  There will be no side effects in other forms. This component uses `DynamicFormLayout` instead of `DynamicFormControlLayout` and I'm guessing thats why the `form-row` class is not getting added. BTW, this same issue exists on other forms. See the form for creating a new metadata field for one example. But it's probably less of a usability issue on smaller forms.
- Hints are now visible. But that makes this component's implementation different from other forms on the site because it's using the `ds-` dynamic-form components directly. Since hints do not appear in other forms, I suspect there's something amiss in the shared form builders. It would be great if you could comment on this, @atarix83 . It won't be a problem to revert  back to the version without visible hints if that's your recommendation. But it's really nice to have them.

mspalti commented:
@tdonohue I changed the scss. The extra spacing should be gone.

Before this is merged it would good to see what @atarix83 thinks about my changes to the form layout.  The problem with hints and spacing comes from the dynamic forms implementation. It might be that correcting individual forms like this one isn't the best approach. Or maybe this is best...

tdonohue reviewed (approved):
👍 Thanks @mspalti ! I've re-tested today, and all my prior feedback has been addressed.

That said, I agree that we should wait to merge this until @atarix83 has a chance to look at the code, since he'd be better able to respond to your questions regarding the dynamic forms implementation.  @atarix83 , if you could get to this soon, I'd appreciate it!

atarix83 reviewed (changes requested):
@tdonohue @mspalti 

I'm sorry for the late review. I think that the implementation works, but it used the wrong approach. @mspalti in your implementation you've add some [form layout properties](https://github.com/DSpace/dspace-angular/pull/1438/files#diff-35ace8b8d29e9653e047da5bb26e942bdfae68aff20877128b2c3b1d2f1cbd75R309-R328) only to hide the iiif fields models. While the right approach is simply to not add at all the iiif fields to [formModels array](https://github.com/DSpace/dspace-angular/pull/1438/files#diff-35ace8b8d29e9653e047da5bb26e942bdfae68aff20877128b2c3b1d2f1cbd75R218) when is not necessary. This means that during the component's init you need to dynamically build the formModel according to which form fields you need. In this way you can rid of the code where you[ change the fields layout properties](https://github.com/DSpace/dspace-angular/pull/1438/files#diff-35ace8b8d29e9653e047da5bb26e942bdfae68aff20877128b2c3b1d2f1cbd75R694-R711)

mspalti commented:
@atarix83 , @tdonohue : thanks for the feedback @atarix83. Your suggestion makes sense. This now adds iiif fields to the form group instead of hiding the ones in the static layout. It make this work I need to initialize the form early and then re-initialize it if the bitstream is iiif enabled. I'm able to switch between different bitstream types and the form renders correctly. 

The metadata updates are all working for me.

atarix83 commented on src/app/bitstream-page/edit-bitstream-page/edit-bitstream-page.component.scss:
```diff
@@ -6,3 +6,7 @@
     }
   }
 }
+:host ::ng-deep ds-dynamic-form-control-container > div > label {
```
is it still necessary ?

mspalti replied:
Yes. I checked this. It is still needed.

atarix83 commented on src/app/bitstream-page/edit-bitstream-page/edit-bitstream-page.component.ts:
```diff
@@ -467,6 +614,35 @@ export class EditBitstreamPageComponent implements OnInit, OnDestroy {
     const primary = rawForm.fileNamePrimaryContainer.primaryBitstream;
     Metadata.setFirstValue(newMetadata, 'dc.title', rawForm.fileNamePrimaryContainer.fileName);
     Metadata.setFirstValue(newMetadata, 'dc.description', rawForm.descriptionContainer.description);
+    console.log(this.isIIIF);
```
this has been unintentionally committed

mspalti replied:
Thanks. I'll clean these up.

atarix83 commented on src/app/bitstream-page/edit-bitstream-page/edit-bitstream-page.component.ts:
```diff
@@ -467,6 +614,35 @@ export class EditBitstreamPageComponent implements OnInit, OnDestroy {
+
+        delete newMetadata[this.IIIF_LABEL_METADATA];
+      } else {
+        console.log(rawForm.iiifLabelContainer.iiifLabel);
```
this has been unintentionally committed

atarix83 reviewed (changes requested):
thanks @mspalti for addressing my feedback, I have only few inline comments and one change request. I would avoid to initialize the form twice ( [here](https://github.com/DSpace/dspace-angular/blob/8bfd6ca03182ea6a977e44d154583801290d7955/src/app/bitstream-page/edit-bitstream-page/edit-bitstream-page.component.ts#L717-L719) and [here](https://github.com/DSpace/dspace-angular/blob/8bfd6ca03182ea6a977e44d154583801290d7955/src/app/bitstream-page/edit-bitstream-page/edit-bitstream-page.component.ts#L427) )  I think that you can do only once after iiif check is done

mspalti commented:
Hi @atarix83 , I thought about initializing the form once. But unfortunately the observable produces no events if the bitstream is not an image or exists outside an IIIF bundle.  Because of that the form is never initialized if I place the initialization only in the subscriber.

mspalti commented:
@atarix83 , I modified this so that the formGroup is initialized only once. Interestingly, I still need to detectChanges in the subscriber to assure that the form always renders completely.

mspalti commented:
In brief, we need to to detectChanges because `formGroup` is not an observable and Angular `async` is not possible for the `*ngIf="formGroup"` that prevents the form from being rendered before `formGroup` is created. Without detectChanges the bitstream edit form sometimes is not added to the page.

tdonohue commented:
@atarix83 : I'm not sure if you've seen @mspalti 's response to your feedback, so I just requested a re-review from you.  If you feel this is still possible to fit into 7.2, there's still time, but we'd need to move quickly, as I'd like to cut 7.2 either tomorrow (Weds) or Thurs if possible.  If you feel this needs more rework, we can figure out an approach to get it into 7.3 instead.

atarix83 reviewed (approved):
thanks @mspalti  for addressing my feedback, using a change detections is not a matter
